aws/resource_aws_athena_named_query.go

```diff
@@ -0,0 +1,93 @@
+package aws
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/athena"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsAthenaNamedQuery() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsAthenaNamedQueryCreate,
+		Read:   resourceAwsAthenaNamedQueryRead,
+		Delete: resourceAwsAthenaNamedQueryDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"query": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"database": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceAwsAthenaNamedQueryCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).athenaconn
+
+	input := &athena.CreateNamedQueryInput{
+		Database:    aws.String(d.Get("database").(string)),
+		Name:        aws.String(d.Get("name").(string)),
+		QueryString: aws.String(d.Get("query").(string)),
+	}
+	if raw, ok := d.GetOk("description"); ok {
+		input.Description = aws.String(raw.(string))
+	}
+
+	resp, err := conn.CreateNamedQuery(input)
+	if err != nil {
+		return err
+	}
+	d.SetId(*resp.NamedQueryId)
+	return resourceAwsAthenaNamedQueryRead(d, meta)
+}
+
+func resourceAwsAthenaNamedQueryRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).athenaconn
+
+	input := &athena.GetNamedQueryInput{
+		NamedQueryId: aws.String(d.Id()),
+	}
+
+	_, err := conn.GetNamedQuery(input)
+	if err != nil {
+		if isAWSErr(err, athena.ErrCodeInvalidRequestException, fmt.Sprintf("Athena Named Query (%s) does not exist", d.Id())) {
```

Comment:
I'm afraid this won't match with the (admittedly odd) error message:

```
2017/11/10 10:12:32 [DEBUG] [aws-sdk-go] DEBUG: Request athena/GetNamedQuery Details:
---[ REQUEST POST-SIGN ]-----------------------------
POST / HTTP/1.1
Host: athena.us-west-2.amazonaws.com
User-Agent: aws-sdk-go/1.12.25 (go1.9.2; darwin; amd64) APN/1.0 HashiCorp/1.0 Terraform/0.11.0-beta1
Content-Length: 55
Authorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20171110/us-west-2/athena/aws4_request, SignedHeaders=content-length;content-type;host;x-amz-date;x-amz-target, Signature=*REDACTED*
Content-Type: application/x-amz-json-1.1
X-Amz-Date: 20171110T101232Z
X-Amz-Target: AmazonAthena.GetNamedQuery
Accept-Encoding: gzip

{"NamedQueryId":"fb403822-2e67-4344-b530-fc488aa79c49"}
-----------------------------------------------------
2017/11/10 10:12:32 [DEBUG] [aws-sdk-go] DEBUG: Response athena/GetNamedQuery Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 400 Bad Request
Connection: close
Content-Length: 181
Content-Type: application/x-amz-json-1.1
Date: Fri, 10 Nov 2017 10:12:32 GMT
X-Amzn-Requestid: aa8710fc-c5ff-11e7-97cf-5fde2fdbf5a4


-----------------------------------------------------
2017/11/10 10:12:32 [DEBUG] [aws-sdk-go] {"__type":"InvalidRequestException","AthenaErrorCode":"NamedQuery: %s does not exists","ErrorCode":"NamedQuery: %s does not exists","Message":"fb403822-2e67-4344-b530-fc488aa79c49"}
```

do you mind loosening up the matching? e.g. just to `"does not exist"`

Reply:
I misunderstood the behaviour of `isAWSErr`. I'll modify. And the same reason why the test didn't pass.